hedera-node/src/main/java/com/hedera/services/txns/crypto/validators/AdjustAllowanceChecks.java

```
@@ -0,0 +1,290 @@
+import com.hederahashgraph.api.proto.java.ResponseCodeEnum;
+import com.hederahashgraph.api.proto.java.TokenAllowance;
+
+import javax.inject.Inject;
+import java.util.ArrayList;
+import java.util.List;
+
+import static com.hedera.services.ledger.properties.NftProperty.OWNER;
+import static com.hedera.services.txns.crypto.helpers.AllowanceHelpers.absolute;
+import static com.hedera.services.txns.crypto.helpers.AllowanceHelpers.hasRepeatedId;
+import static com.hedera.services.txns.crypto.helpers.AllowanceHelpers.hasRepeatedSerials;
+import static com.hedera.services.txns.crypto.helpers.AllowanceHelpers.hasRepeatedSpender;
+import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.AMOUNT_EXCEEDS_TOKEN_MAX_SUPPLY;
+import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.EMPTY_ALLOWANCES;
+import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.FUNGIBLE_TOKEN_IN_NFT_ALLOWANCES;
+import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.INVALID_TOKEN_NFT_SERIAL_NUMBER;
+import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.NEGATIVE_ALLOWANCE_AMOUNT;
+import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.NFT_IN_FUNGIBLE_TOKEN_ALLOWANCES;
+import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.OK;
+import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.REPEATED_SERIAL_NUMS_IN_NFT_ALLOWANCES;
+import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.SENDER_DOES_NOT_OWN_NFT_SERIAL_NO;
+import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.SPENDER_ACCOUNT_REPEATED_IN_ALLOWANCES;
+
+public class AdjustAllowanceChecks implements AbstractAllowanceChecks {
+	protected final TypedTokenStore tokenStore;
+	protected final TransactionalLedger<NftId, NftProperty, MerkleUniqueToken> nftsLedger;
+	protected final GlobalDynamicProperties dynamicProperties;
+
+	@Inject
+	public AdjustAllowanceChecks(
+			final TransactionalLedger<NftId, NftProperty, MerkleUniqueToken> nftsLedger,
```

**Especially** since this code can currently run in precheck, I would suggest just injecting the `Supplier<MerkleMap<EntityNumPair, MerkleUniqueToken>>` here, since the `TransactionalLedger` implementations is more expensive and dangerous to use in almost every way.  😁

---

Ah got it ! Changed to use `Supplier<MerkleMap<EntityNumPair, MerkleUniqueToken>>`.  Thanks !

Agreed about the discussion. I thought doing these checks in precheck helps to fail fast.